wrapper/src/main/java/software/amazon/jdbc/plugin/efm/MonitorImpl.java

```diff
@@ -137,125 +140,142 @@ public void clearContexts() {
+            if (firstAddedNewMonitorContext == newMonitorContext) {
+              // This context has already been processed.
+              // Add it back to the queue and process it in the next round.
               this.newContexts.add(newMonitorContext);
-              if (firstAddedNewMonitorContext == null) {
-                firstAddedNewMonitorContext = newMonitorContext;
+              break;
+            }
+            if (newMonitorContext.isActiveContext()) {
+              if (newMonitorContext.getExpectedActiveMonitoringStartTimeNano() > currentTimeNano) {
+                // The context active monitoring time hasn't come.
+                // Add the context to the queue and check it later.
+                this.newContexts.add(newMonitorContext);
+                if (firstAddedNewMonitorContext == null) {
+                  firstAddedNewMonitorContext = newMonitorContext;
+                }
+              } else {
+                // It's time to start actively monitor this context.
+                this.activeContexts.add(newMonitorContext);
               }
-            } else {
-              // It's time to start actively monitor this context.
-              this.activeContexts.add(newMonitorContext);
             }
           }
-        }
-
-        if (!this.activeContexts.isEmpty()) {
 
-          final long statusCheckStartTimeNano = this.getCurrentTimeNano();
-          this.contextLastUsedTimestampNano = statusCheckStartTimeNano;
+          if (!this.activeContexts.isEmpty()) {
 
-          final ConnectionStatus status =
-              checkConnectionStatus(this.nodeCheckTimeoutMillis);
+            final long statusCheckStartTimeNano = this.getCurrentTimeNano();
+            this.contextLastUsedTimestampNano = statusCheckStartTimeNano;
 
-          long delayMillis = -1;
-          MonitorConnectionContext monitorContext;
-          MonitorConnectionContext firstAddedMonitorContext = null;
+            final ConnectionStatus status =
+                checkConnectionStatus(this.nodeCheckTimeoutMillis);
 
-          while ((monitorContext = this.activeContexts.poll()) != null) {
+            long delayMillis = -1;
+            MonitorConnectionContext monitorContext;
+            MonitorConnectionContext firstAddedMonitorContext = null;
 
-            synchronized (monitorContext) {
-              // If context is already invalid, just skip it
-              if (!monitorContext.isActiveContext()) {
-                continue;
-              }
+            while ((monitorContext = this.activeContexts.poll()) != null) {
 
-              if (firstAddedMonitorContext == monitorContext) {
-                // this context has already been processed by this loop
-                // add it to the queue and exit this loop
-                this.activeContexts.add(monitorContext);
-                break;
-              }
+              synchronized (monitorContext) {
```

Comment:
Does this really have to be synchronized? It is a local variable

Reply:
It's a local variable but we add it to queues in MonitorImpl that runs in a separate thread. Monitoring thread may change an internal state of the context. The idea was to synchronize on it to avoid multi-threading collisions. It seems this intent isn't properly implemented here. Need to address it in a separate PR.

Reply:
After some investigation, it seems like the current implementation is working and a fix isn't required.